Add sample application

PeterKneale opened the pull request:
Added a sample application consisting of three projects
- a web api that publishes from a controller and subscribes via a background service
- a console application that subscribes via a background service and publishes within an event handler
- a shared models assembly containing event dto's

- Further work required:
  - Demonstrate use of `PublishMetaData`
  - Demonstrate integration with [CorrelationId](https://www.nuget.org/packages/CorrelationId/)
  - Demonstrate use of Cancellation Tokens
  - Add acceptance tests project
  - Add dockerfiles for api, console, aws and tests
  - Demonstrate docker-compose executing acceptance tests against api / console while using mocked aws environment

relates to issue https://github.com/justeat/JustSaying/issues/476

PeterKneale commented on JustSaying.Sample.Restaurant.OrderingApi/Handlers/OrderReadyEventHandler.cs:
```diff
@@ -0,0 +1,24 @@
+        public Task<bool> Handle(OrderReadyEvent message)
+        {
+            _log.LogInformation($"*** Order {message.OrderId} Ready! ***");
+
```
Comment on what we might do here? ie: mark order as ready in websites database

PeterKneale commented on JustSaying.Sample.Restaurant.OrderingApi/Models/CustomerOrderModel.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public class CustomerOrderModel
+    {
+        public string Description { get; set; }
```
add ctor and make sets private

PeterKneale commented on JustSaying.Sample.Restaurant.OrderingApi/Program.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Design", "CA1052:Static holder types should be Static or NotInheritable")]
```
should `CodeAnalysis` be enabled for sample code?

PeterKneale commented on JustSaying.Sample.Restaurant.OrderingApi/Subscriber.cs:
```diff
@@ -0,0 +1,28 @@
+        {
+            _log.LogInformation("*** Ordering API Subscriber Running ***");
+
+            _bus.Start(stoppingToken);
```
is this correct?

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+                      });
+                      config.Messaging(x =>
+                      {
+                          x.WithRegion(RegionEndpoint.APSoutheast2);
```
Make this a config value? For example, we primarily use EU West 1.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+                      });
+                  });
+                  services.AddJustSayingHandler<OrderPlacedEvent, OrderPlacedEventHandler>();
+                  services.AddHostedService<Subscriber>();
```
I think that's right - in fact I've had the idea of creating a built-in hosted service for JustSaying (that would be a new NuGet package) that would do some of this boiler plate to make it even easier to do something like this. We have a lot of worker services we use to process messages, and as we move more of them to .NET Core over time, it would be great to reduce as much repetitive boiler plate code for such use cases as possible.

For example `services.AddJustSayingService<MyService>()` where there could be a base class that could be derived from to do configuration, and then everything else gets bootstrapped up.

Of course that's a totally separate thing, but if we get this nicely working in the sample it would be a good jumping off point to investigate doing something like that.

PeterKneale replied:
One thing i wanted to explore further was building a cron scheduled background task then integrating that with JustSaying so that I could publish a message on a schedule. My thinking was it would help decouple what a service does from how and when it is triggered
`*/10 * * * *` -> `OrderExpiryCheckRequiredEvent`
`0 */4 * * *` -> `OrderCancellationSummaryRequiredEvent`

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+{
+    internal class Program
+    {
+        private static async Task Main()
```
Nit: `private` seems odd to me for the main method of an app.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/OrderPlacedEventHandler.cs:
```diff
@@ -0,0 +1,37 @@
+using JustSaying.Messaging.MessageHandling;
+using JustSaying.Sample.Restaurant.Models;
+using Microsoft.Extensions.Logging;
+using System.Threading.Tasks;
```
Nit: sort using statements where System directives take precedence over the others. I guess we should find a way to editor-config that as the default.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/appsettings.json:
```diff
@@ -0,0 +1,10 @@
+      "Microsoft": "Warning"
+    }
+  },
+  "AllowedHosts": "*"
```
I guess this setting is redundant for a console app?

PeterKneale replied:
yeah it was a copy/paste error - done

martincostello commented on JustSaying.Sample.Restaurant.Models/OrderPlacedEvent.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace JustSaying.Sample.Restaurant.Models
+{
+    public class OrderPlacedEvent : Message
```
Simple POCOs with public gets and sets and no constructor are fine.

slang25 replied:
Agreed, a private constructor could trip up the deserialization

PeterKneale replied:
Isn't an immutable type most appropriate so that the subscriber is unable to change any component of the message that was received (intentionally or otherwise).
Also the publisher is only able to construct messages in a valid fashion via the constructor.
(The serializer is fine with this type)
```cs
    public class OrderPlacedEvent : Message
    {
        public OrderPlacedEvent(int orderId, string description)
        {
            if (orderId <= 0)
            {
                throw new ArgumentException("orderId must be positive", nameof(orderId));
            }
            if (string.IsNullOrWhiteSpace(description))
            {
                throw new ArgumentException("description must be supplied", nameof(description));
            }

            OrderId = orderId;
            Description = description;
        }

        public int OrderId { get; private set; }

        public string Description { get; private set; }
    }
```

martincostello replied:
We've had issues in the past where trying to force business rules (what's required, what's optional etc.) into the POCOs, which are ultimately for serialization in the world of JustSaying, have caused problems as applications have evolved over time.

The biggest one being when you need to add a new optional field, so you add an extra constructor overload that sets it so it's not a binary breaking change, but forget to add appropriate hints to the serializer. Then you get runtime errors because the serializer doesn't know which constructor to use.

It also makes it more verbose when you're writing tests.

Ideally in a production application, validation etc. is handled elsewhere, not in the internals of JustSaying as a consequence of message publishing of deserialization.

PeterKneale replied:
Ah, understood - Thanks!

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/JustSaying.Sample.Restaurant.OrderingApi.csproj:
```diff
@@ -0,0 +1,23 @@
+    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
+    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+    <UserSecretsId>27d3f468-c0a6-4791-80a3-4ce69e75393f</UserSecretsId>
```
Set this to a name, rather than a GUID.

PeterKneale replied:
ta - never used this before.

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Program.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Design", "CA1052:Static holder types should be Static or NotInheritable")]
```
I'd say yes by default, but you could turn off this specific one by adding `<NoWarn>$(NoWarn);CA1052</NoWarn>` to the `.csproj` file.

martincostello replied:
Actually, this is easy to fix - just make it static.

PeterKneale replied:
ahh - ta - done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -0,0 +1,86 @@
+                });
+                config.Messaging(x =>
+                {
+                    x.WithRegion(RegionEndpoint.APSoutheast2);
```
Config.

martincostello commented on README.md:
```diff
@@ -284,6 +284,31 @@ At this point, the power tool is only able to move an arbitrary number of messag
 
+## Sample Application
+
+- To run the sample application against a simulated aws sqs / sns endpoint run this container
```
Nit: _"...AWS SQS/SNS..."_

PeterKneale replied:
done

martincostello commented on README.md:
```diff
@@ -284,6 +284,31 @@ At this point, the power tool is only able to move an arbitrary number of messag
+- Alternatively to use your real AWS account
+    - Locate the setup code `services.AddJustSaying(...)` in both `Program.cs` and `Startup.cs`
+    - Remove the references to the 'ServiceUrl`
+    - Add references to `x.WithCredentials(...)` supplying your real aws credentials
```
Nit: _"AWS"_

PeterKneale replied:
done

martincostello commented on README.md:
```diff
@@ -284,6 +284,31 @@ At this point, the power tool is only able to move an arbitrary number of messag
+
+- Demonstrates
+  - Publishing messages to a SNS Topic from a WebApi application and Console Application
+  - Receiving messages from a SQS queue subscribed to a SNS topic in a WebApi application and Console Application
```
_"WebApi"_ -> _"ASP.NET Core web"_

PeterKneale replied:
done

slang25 commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+              })
+              .UseConsoleLifetime()
+              .Build()
+              .RunAsync().ConfigureAwait(false);
```
No need for `ConfigureAwait(false)` in this project. Console apps and ASP.NET Core applications do not have a `SyncronizationContext`.

martincostello replied:
I think this is more a reaction to the catch-all code analyser suggestions.

martincostello replied:
I think this is more a reaction to the catch-all code analyser suggestions.

PeterKneale replied:
It is - I've had to add a `SuppressMessage` in order to make it compile without `ConfigureAwait`

slang25 commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+                      });
+                  });
+                  services.AddJustSayingHandler<OrderPlacedEvent, OrderPlacedEventHandler>();
+                  services.AddHostedService<Subscriber>();
```
This is exactly what I had in mind 👍

slang25 commented on JustSaying.Sample.Restaurant.OrderingApi/Controllers/OrdersController.cs:
```diff
@@ -0,0 +1,44 @@
+                Description = order.Description
+            };
+
+            await _publisher.PublishAsync(message).ConfigureAwait(false);
```
again remove `ConfigureAwait` please

PeterKneale commented on JustSaying.Sample.Restaurant.KitchenConsole/OrderPlacedEventHandler.cs:
```diff
@@ -0,0 +1,37 @@
+            };
+
+            await _publisher.PublishAsync(orderReadyEvent).ConfigureAwait(false);
+            return true;
```
done

PeterKneale commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+                      });
+                      config.Subscriptions(x =>
+                      {
+                          x.ForTopic<OrderPlacedEvent>();
```
done

PeterKneale commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,55 @@
+                      });
+                      config.Publications(x =>
+                      {
+                          x.WithTopic<OrderReadyEvent>();
```
done

PeterKneale commented:
- Have pushed, can squash if required.
- Have had to add suppression in order to remove the `ConfigureAwait`
`[assembly: SuppressMessage("Reliability", "CA2007:Do not directly await a Task", Justification = "Console application do not have a SynchronizationContext")]`

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -1,55 +1,90 @@
 using Microsoft.Extensions.Logging;
-using System;
-using System.Threading.Tasks;
+[assembly: SuppressMessage("Reliability", "CA2007:Do not directly await a Task", Justification = "Console application do not have a SynchronizationContext")]
```
I think it would be best to use `<NoWarn>$(NoWarn);CA2007</NoWarn>` to do this in the project file instead.

Otherwise these assembly-wide scoped suppression are hard to find to the code files.

PeterKneale replied:
ok - done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -1,55 +1,90 @@
             await new HostBuilder()
+                .ConfigureAppConfiguration((hostContext, config) =>
+                {
+                    config.SetBasePath(Environment.CurrentDirectory);
```
This wouldn't be correct if you ran `dotnet` with an absolute path from outside the directory tree. I can't remember the exact name, but I think you want `ApplicationContext.ApplicationBase` or something like that instead.

PeterKneale replied:
ok. Taking a look

PeterKneale replied:
You're right. `AppContext.BaseDirectory` allows it to be executed from another path

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -1,55 +1,90 @@
+                           else
+                           {
+                               // The real AWS environment will require some means of authentication
+                               //x.WithBasicCredentials("###", "###");
```
If you add _UserSecrets_ to the config providers, you could load these from config without needing to have it commented-out with placeholders.

PeterKneale replied:
Yeah im a bit torn - im trying to be careful to leave the sample app in the simplest possible state where people can easily see what they might need to do or change in order to get it running on their box. Some have keys, others env variables etc. Not sure whats best.
Can go with the secrets if you think that's best but I think its a bit of a distraction from demonstrating the bus.

PeterKneale replied:
on a similar note i was thinking of leaving a line like this for those who configure the region via environment variables too..?
```cs
//x.WithRegion(new Amazon.Runtime.EnvironmentVariableAWSRegion().Region);
```

martincostello replied:
I guess that's more of an advanced use case? We wouldn't want to balloon the samples too much with all the possible things you _could_ do. Also if you were using environment variables, you'd probably be using the .NET provider for that and binding them to `IConfiguration` somehow I'd have thought.

PeterKneale replied:
Yeah and it feels a bit orthogonal to the actual demonstration of your library. Its more demonstrating aspnet core configuration and the aws sdk

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/JustSaying.Sample.Restaurant.OrderingApi.csproj:
```diff
@@ -20,4 +23,10 @@
   </ItemGroup>
 
+  <ItemGroup>
+    <Content Update="appsettings.json">
```
Could you simplify this to this form?

```xml
<Content Update="appsettings.json" CopyToOutputDirectory="PreserveNewest" />
```

PeterKneale replied:
Done - didnt know the elements could be converted to properties like that

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Program.cs:
```diff
@@ -1,14 +1,12 @@
-[assembly: SuppressMessage("Performance", "CA1822:Mark members as static")]
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+[assembly: SuppressMessage("Reliability", "CA2007:Do not directly await a Task", Justification = "Console application do not have a SynchronizationContext")]
```
Use `<NoWarn>`.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -28,26 +30,46 @@ public void ConfigureServices(IServiceCollection services)
-                    //x.WithCredentials(...)
+                    if (!string.IsNullOrWhiteSpace(_awsServiceUrl))
+                    {
+                        // The AWS client sdk allows specifying a custom http endpoint.
```
Nit: _"SDK"_

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -28,26 +30,46 @@ public void ConfigureServices(IServiceCollection services)
-                    //x.WithCredentials(...)
+                    if (!string.IsNullOrWhiteSpace(_awsServiceUrl))
+                    {
+                        // The AWS client sdk allows specifying a custom http endpoint.
```
Nit: _"HTTP"_

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -28,26 +30,46 @@ public void ConfigureServices(IServiceCollection services)
+                    else
+                    {
+                        // The real AWS environment will require some means of authentication
+                        //x.WithBasicCredentials("###", "###");
```
As above regarding user secrets.

AnthonySteele commented on JustSaying.Sample.Restaurant.OrderingApi/Controllers/OrdersController.cs:
```diff
@@ -0,0 +1,44 @@
+
+            await _publisher.PublishAsync(message);
+
+            _log.LogInformation($"*** Order {orderId} Placed ***");
```
As a best practice, I would remove the `*` characters, and structure the log entry, i.e.

` _log.LogInformation("Order {orderId} Placed", orderId);`

AnthonySteele replied:
And a few other places too, I think.

PeterKneale replied:
Yes, all done.

AnthonySteele commented on JustSaying.Sample.Restaurant.OrderingApi/Subscriber.cs:
```diff
@@ -0,0 +1,32 @@
+
+        protected override Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            _log.LogInformation("*** Ordering API Subscriber Running ***");
```
Logging as above

PeterKneale commented:
Further feedback welcome

martincostello commented:
@PeterKneale Thanks for your work on this. When I get some time today or tomorrow, I'll pull your branch locally and have a play around with what's there so far. 👍

PeterKneale commented:
Sure thing. Thanks for a great library

martincostello commented on README.md:
````diff
@@ -284,6 +284,31 @@ At this point, the power tool is only able to move an arbitrary number of messag
+    docker run -d --name goaws -p 4100:4100 pafortin/goaws
+    ```
+- Alternatively to use your real AWS account
+    - Locate the setup code `services.AddJustSaying(...)` in both `Program.cs` and `Startup.cs`
````
At a later point, we should link these directly to the relevant code using permalinks.

PeterKneale replied:
Done. I've just linked to the files, not the lines in question as I find that a bit to subject to breaking as code churns

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/OrderPlacedEventHandler.cs:
```diff
@@ -0,0 +1,49 @@
+    public class OrderPlacedEventHandler : IHandlerAsync<OrderPlacedEvent>
+    {
+        private readonly IMessagePublisher _publisher;
+        private readonly ILogger<OrderPlacedEventHandler> _log;
```
Rename to `_logger`.

PeterKneale replied:
fixed

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/OrderPlacedEventHandler.cs:
```diff
@@ -0,0 +1,49 @@
+
+            _log.LogInformation("Order {orderId} ready", message.OrderId);
+
+            var orderReadyEvent = new OrderReadyEvent
```
Wrap the "logic" in a try-catch, logging the exception in the catch block and returning false, e.g.:

```csharp
catch (Exception ex)
{
    _logger.LogError(ex, "Failed to handle message for {orderId}", message.OrderId);
    return false;
}
````

PeterKneale replied:
fixed

PeterKneale replied:
I rebased on master and get:
```
error CA1031: Modify 'Handle' to catch a more specific exception type, or rethrow the exception.
```

I've suppressed it via
```xml
<NoWarn>.............;CA1031</NoWarn>
```

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,88 @@
+                    config.AddJsonFile($"appsettings.{hostContext.HostingEnvironment.EnvironmentName}.json", optional: true);
+                    config.AddEnvironmentVariables();
+                })
+               .ConfigureLogging((hostContext, logging) =>
```
Could simplify this to:

```csharp
.ConfigureLogging(loggingBuilder => loggingBuilder.AddConsole())
```

PeterKneale replied:
fixed

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,88 @@
+               })
+               .ConfigureServices((hostContext, services) =>
+               {
+                   var awsRegion = hostContext.Configuration.GetValue<string>("AWSRegion");
```
Could just use `hostContext.Configuration["AWSRegion"];` etc. here as we're just using strings.

PeterKneale replied:
fixed

martincostello commented on JustSaying.Sample.Restaurant.Models/JustSaying.Sample.Restaurant.Models.csproj:
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
```
For models can target `netstandard2.0` instead.

PeterKneale replied:
fixed

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/JustSaying.Sample.Restaurant.OrderingApi.csproj:
```diff
@@ -0,0 +1,31 @@
+    <PackageReference Include="Microsoft.AspNetCore.Razor.Design" Version="2.2.0" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.2.0" />
+    <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.0.2105168" />
```
Remove this reference.

PeterKneale replied:
fixed

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Program.cs:
```diff
@@ -0,0 +1,22 @@
+        public static IWebHostBuilder CreateWebHostBuilder(string[] args)
+        {
+            return WebHost.CreateDefaultBuilder(args)
+                .UseStartup<Startup>();
```
Add `.ConfigureLogging((loggingBuilder) => loggingBuilder.AddConsole())`

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -0,0 +1,109 @@
+{
+    public class Startup
+    {
+        private readonly string _awsRegion;
```
Access these inline from `IConfiguration` in the builder calls using the overload that receives an `IServiceProvider`, rather than storing here in this class. That would then make it easier to move the builder code out into another class with some extension methods without anchoring it to the `Startup` class.

PeterKneale replied:
Sorry not too sure what you mean here. 
I've created extension methods for `IConfiguration` so that this startup code can be moved / refactored and the configuration keys are in only one place.

martincostello replied:
I was thinking something like this:

```csharp
services.AddJustSaying((config, serviceProvider) =>
{
    var configuration = serviceProvider.GetRequiredService<IConfiguration>();

    // ...

    config.Messaging(x =>
    {
        x.WithRegion(configuration.GetAWSRegion());
    });

    // ...
}
```

This was initially because I'd misremembered and thought that `IConfiguration` wasn't going to be injectable into `Startup` anymore in 3.0, but it appears that either I was wrong about that, or it's been added back (https://github.com/aspnet/Announcements/issues/353) since an earlier announcement of the changes.

PeterKneale replied:
ok. shall i leave as is then?

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -0,0 +1,109 @@
+            // Add a background service that is listening for messages related to the above subscriptions
+            services.AddHostedService<Subscriber>();
+
+            services.AddLogging(c =>
```
Remove, with logging configured in `Program` instead.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -0,0 +1,109 @@
+                c.AddConsole();
+            });
+
+            services.AddSwaggerGen(c =>
```
If we had a simple Bootstrap UI, we could remove the need to use Swashbuckle at all.

PeterKneale replied:
See comments below

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -0,0 +1,109 @@
+            }
+            else
+            {
+                app.UseHsts();
```
For simplicity regarding HTTPS configuration, remove the HTTPS and HSTS method calls.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Subscriber.cs:
```diff
@@ -0,0 +1,32 @@
+    public class Subscriber : BackgroundService
+    {
+        private readonly IMessagingBus _bus;
+        private readonly ILogger<Subscriber> _log;
```
`_logger`

martincostello commented:
I've cloned the the sample locally and had a play around with it. Here's a few bits of feedback on specifics, but overall this is pretty good.

I think rather than Swagger and an "API", we should have a simple one-page MVC app with Bootstrap added, then when you run the sample it can present a UI where the order can be picked from a few options (e.g. _pizza_, _curry_, _burger_, etc.) with a few simple JavaScript calls to post the order and poll for and update (simple meta refresh?) when the "kitchen" says it is ready.

This is more natural to navigate that doing a manual HTTP POST in the Swagger UI and provides feedback that's a bit more visual that watching two consoles, though that's still there for a more "low level" view.

I'd also add some additional handlers that simulates payment and store the order in an in-memory database in the web app to persist things. Then if messages fail the process can be potentially recovered by leveraging the reliability of building on top of a messaging system. This would also demonstrate things like have multiple subscriptions to the same queue for different purposes.

Then we can flesh out the sample with a few extra handlers/processes, something like:

  * User creates order in UI via POST to back-end
  * UI back-end publishes "placed" message to queue
  * UI back-end receives "placed" message and stores in in-memory database
  * User "pays" for the order in the UI via POST to back-end
  * UI back-end publishes "paid" message to topic
  * UI back-end receives "paid" message and updates order in in-memory database with new state
  * UI updates to show the user payment made and restaurant is awaiting confirmation
  * Separate handler receives "paid" message and sends order to restaurant queue.
  * Restaurant handler receives "order" message and sends "accepted" message to topic
  * UI back-end receives "accepted" message and updates order in in-memory database with new state
  * UI updates to show the user the order has been "accepted" by the restaurant to be cooked

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,88 @@
+                               // The AWS client SDK allows specifying a custom HTTP endpoint.
+                               // For testing purposes it is useful to specify a value that
+                               // points to a docker image such as `p4tin/goaws` or `localstack/localstack`
+                               x.WithServiceUrl(awsServiceUrl);
```
Add a call to `WithAnonymousCredentials()` (see #528).

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Startup.cs:
```diff
@@ -0,0 +1,109 @@
+                        // The AWS client SDK allows specifying a custom HTTP endpoint.
+                        // For testing purposes it is useful to specify a value that
+                        // points to a docker image such as `p4tin/goaws` or `localstack/localstack`
+                        x.WithServiceUrl(_awsServiceUrl);
```
Add a call to `WithAnonymousCredentials()` (see #528).

PeterKneale replied:
done

PeterKneale commented:
I've addressed your basic feedback. I like the idea about building the web frontend and demonstrating a more thorough workflow.. I'm wondering what delivers the most advantage because I've got limited time.

An area that I think also really needs to be added is an example of how to build dockerised system tests. 
ie: 
- A docker compose file/s consisting of
  - p4tin/goaws
  - sql
  - OrderingApi
  - KitchenConsole
  - XUnit System Tests

- The system tests can publish an event, then patiently await an expected outcome in the database for a few seconds. 
- The tests could perform a HTTP POST to the API, then patiently await an expected outcome in the database for a few seconds. 
- I'm doing this at work at the moment and it feels like one of the only ways to test the interaction of multiple services but would welcome any other ideas.

martincostello commented:
I think it's fine to keep the scope more confined here, than my original suggestion/comments. That way this delivers value into master sooner, and then it can be iterated on over time to have the option of a more "interactive" sample as well as the lower-level one here.

Otherwise what you've alluded to above sounds reasonable. Often the acceptance/integration/system tests for our own workers that use JustSaying involve observing the side effects of the implementation details for message publishing and consumption (e.g. a record appears in a database).

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/ConfigurationExtensions.cs:
```diff
@@ -0,0 +1,28 @@
+            return !string.IsNullOrWhiteSpace(configuration[AWSServiceUrlKey]);
+        }
+
+        public static Uri GetAWSServiceUrl(this IConfiguration configuration)
```
Nit: As this is a `Uri`, could you rename to `GetAWSServiceUri()`?

PeterKneale replied:
done. I went back and forth on this because the underlying config was `...ServiceUrl`

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/JustSaying.Sample.Restaurant.KitchenConsole.csproj:
```diff
@@ -0,0 +1,29 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <NoWarn>$(NoWarn);CA2007;CA1031</NoWarn>
```
Nit: warnings in ascending numerical order.

PeterKneale replied:
done

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/OrderPlacedEventHandler.cs:
```diff
@@ -0,0 +1,58 @@
+        {
+            _logger.LogInformation("Order {orderId} for {description} received", message.OrderId, message.Description);
+
+            // TODO: handle the actual order
```
Is this a `TODO` for you, or a note to the reader? If the latter, I'd rephrase into commentary, as otherwise it potentially makes the sample look unfinished.

PeterKneale replied:
good point - done

martincostello commented on JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:
```diff
@@ -0,0 +1,84 @@
+            await new HostBuilder()
+                .ConfigureAppConfiguration((hostContext, config) =>
+                {
+                    config.SetBasePath(AppContext.BaseDirectory);
```
Is there a method on `HostBuilder` anywhere that does all this by default? `WebHostBuilder` does similar (or comes preconfigured that way), so would shave off a few lines of code if it's not needed or there's a one-liner.

PeterKneale replied:
Just removing it seems to work...
executing `dotnet run` from the projects folder works as does ` dotnet run --project samples/JustSaying.Sample.Restaurant.KitchenConsole/JustSaying.Sample.Restaurant.KitchenConsole.
csproj`

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/ConfigurationExtensions.cs:
```diff
@@ -0,0 +1,28 @@
+            return !string.IsNullOrWhiteSpace(configuration[AWSServiceUrlKey]);
+        }
+
+        public static Uri GetAWSServiceUrl(this IConfiguration configuration)
```
Nit: `GetAWSServiceUri()`.

PeterKneale replied:
done

martincostello commented on JustSaying.Sample.Restaurant.OrderingApi/Properties/launchSettings.json:
```diff
@@ -0,0 +1,35 @@
+    "IIS Express": {
+      "commandName": "IISExpress",
+      "launchBrowser": true,
+      "launchUrl": "",
```
As you've used in-process hosting, you can specify `"ancmHostingModel": "InProcess"` here to have IIS Express match.

PeterKneale replied:
done

martincostello commented:
@PeterKneale Just some minor comments around naming and comments, otherwise this looks really good.

We're planning on moving the code around soon to tidy up the mess of folders that's arisen over time into a `src`, `tests` structure as has become a .NET Core "standard" recently. As an initial step towards that, could you move all the new things into a `samples` folder in the repo root please, instead of putting them directly at the root? That'll save the extra churn later when we shift things around.

Again, thanks for all your work on this! 👍

PeterKneale commented:
- moved to samples folder
- addressed feedback
- squashed & pushed
- will add bootstrap skeleton and remove swashbuckle in another PR later tonight if i get a chance and start iterating towards a more complete sample over time

PeterKneale commented on samples/JustSaying.Sample.Restaurant.KitchenConsole/appsettings.json:
```diff
@@ -0,0 +1,11 @@
+  "AWSRegion": "eu-west-1",
+  "AWSServiceUrl": "http://localhost:4100",
+  "Logging": {
+    "LogLevel": {
```
Should I be able to add `"EventLog": "Trace"` here and have the messages contents logged?

martincostello replied:
I think so - at some point in v7 we need to rename that, because it's easy to confuse with the Windows Event Log.

martincostello commented:
I think I'm happy with this as it is as a starting point. If the rest of the maintainers are happy, I think we could get this merged next week and further changes can be done in follow-up PRs over time.

PeterKneale commented:
👍

martincostello commented:
@PeterKneale Is there any more you're planning to do as part of this PR to justify the WIP status still, or should we merge this?

PeterKneale commented:
Ready to merge - I cant remove the WIP label.

martincostello commented:
Excellent work @PeterKneale! :shipit:
